docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -51,10 +51,43 @@ The current multisig public keys use amino serialization to generate the address
 those public keys and their address formatting, and call them "legacy amino" multisig public keys
 in protobuf. We will also create multisig public keys without amino addresses to be described below.
 
-### Canonical Public Key Addresses
 
-Following on the discussion in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we propose the
-following approach.
+### Canonical Address Format
+
+We have three types of accounts we would like to create addresses for in the future:
+- regular public key addresses for new signature algorithms (ex. `sr25519`).
+- public key addresses for multisig public keys that don't use amino encoding
+- module accounts: basically any accounts which cannot sign transactions and
+which are managed internally by modules
+
+To address all of these use cases we propose the following basic `AddressHash` function,
+based on the discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694):
+
+```go
+func AddressHash(prefix string, contents []byte) []byte {
+	preImage := []byte(prefix)
+	if len(contents) != 0 {
+		preImage = append(preImage, 0)
+		preImage = append(preImage, contents...)
+	}
```

Comment:
When implementing this we should use use `make([]byte, correctSize)` and copy they content using `for` or `copy` (into slice starting from index `2`).